sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(
+          &tm.token, &out_response->registration_information.error_tracking_id));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("message"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(
+          &tm.token, &out_response->registration_information.error_message));
+    }
+    else if (az_span_is_content_equal(AZ_SPAN_FROM_STR("timestampUtc"), tm.name))
+    {
+      AZ_RETURN_IF_FAILED(az_json_token_get_string(
+          &tm.token, &out_response->registration_information.error_timestamp));
+    }
+    else if (tm.token.kind == AZ_JSON_TOKEN_OBJECT_START)
+    {
+      AZ_RETURN_IF_FAILED(az_json_parser_skip_children(&jp, tm.token));
+    }
+    else if (az_succeeded(az_iot_provisioning_client_payload_error_code_parse(
+                 &tm, &out_response->registration_information)))
+    {
+      found_error = true;
+    }
+
+    // else ignore token.
+  }
+
+  if (!(found_registration_state && found_operation_id))
+  {
+    out_response->operation_id = AZ_SPAN_NULL;
+    out_response->registration_state = AZ_SPAN_FROM_STR("failed");
+
+    if (!found_error)
+    {
+      return AZ_ERROR_ITEM_NOT_FOUND;
+    }
+  }
+
+  return AZ_OK;
+}
+
+/*
+Example flow:
+
+Stage 1:
+ topic: $dps/registrations/res/202/?$rid=1&retry-after=3
+ payload:
+  {"operationId":"4.d0a671905ea5b2c8.e7173b7b-0e54-4aa0-9d20-aeb1b89e6c7d","status":"assigning"}
+
+Stage 2:
+  {"operationId":"4.d0a671905ea5b2c8.e7173b7b-0e54-4aa0-9d20-aeb1b89e6c7d","status":"assigning",
+  "registrationState":{"registrationId":"paho-sample-device1","status":"assigning"}}
+
+Stage 3:
+ topic: $dps/registrations/res/200/?$rid=1
+ payload:
+  {"operationId":"4.d0a671905ea5b2c8.e7173b7b-0e54-4aa0-9d20-aeb1b89e6c7d","status":"assigned",
+  "registrationState":{ ... }}
+
+ Error:
+ topic: $dps/registrations/res/401/?$rid=1
+ payload:
+ {"errorCode":401002,"trackingId":"8ad0463c-6427-4479-9dfa-3e8bb7003e9b","message":"Invalid
+  certificate.","timestampUtc":"2020-04-10T05:24:22.4718526Z"}
+*/
+AZ_NODISCARD az_result az_iot_provisioning_client_parse_received_topic_and_payload(
     az_iot_provisioning_client const* client,
     az_span received_topic,
     az_span received_payload,
     az_iot_provisioning_client_register_response* out_response)
 {
   (void)client;
-  (void)received_topic;
-  (void)received_payload;
-  (void)out_response;
 
-  return AZ_ERROR_NOT_IMPLEMENTED;
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(received_topic, 0, false);
+  AZ_PRECONDITION_VALID_SPAN(received_payload, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_response);
+
+  int32_t idx = az_span_find(received_topic, str_dps_registrations_res);
+  if (idx != 0)
+  {
+    return AZ_ERROR_IOT_TOPIC_NO_MATCH;
+  }
+
+  // Parse the status.
+  az_span remainder = az_span_slice(
+      received_topic, az_span_size(str_dps_registrations_res), az_span_size(received_topic));
```

Comment:
Same here/elsewhere. Use `az_span_slice_to_end`
```suggestion
  az_span remainder = az_span_slice_to_end(received_topic, az_span_size(str_dps_registrations_res));
```

Reply:
I can't find other places - let me know if you saw another.

Reply:
Nope, no other places found.